Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -1636,37 +1900,38 @@ static bool isExternalTexture(const AST::Variable& variable)
     return bindingTypeForType(variable.storeType()) == BindingType::TextureExternal;
 }
 
-static bool typesMatch(const AST::Variable& variable, const BindGroupLayoutEntry::BindingMember& bindingMember)
+static std::optional<String> errorValidatingTypes(const AST::Variable& variable, const BindGroupLayoutEntry::BindingMember& bindingMember)
 {
-    return WTF::switchOn(bindingMember, [&](const BufferBindingLayout&) {
-        return isBuffer(variable);
-    }, [&](const SamplerBindingLayout& samplerBinding) {
-        return isSampler(variable, samplerBinding.type);
-    }, [&](const TextureBindingLayout& textureBinding) {
-        return isTexture(variable, textureBinding.multisampled);
-    }, [&](const StorageTextureBindingLayout& storageTexture) {
-        return isStorageTexture(variable, storageTexture.access);
-    }, [&](const ExternalTextureBindingLayout&) {
-        return isExternalTexture(variable);
+    using ResultType = std::optional<String>;
+    return WTF::switchOn(bindingMember, [&](const BufferBindingLayout&) -> ResultType {
+        return isBuffer(variable) ? ResultType { std::nullopt } : ResultType { "WGSL variable is not a buffer"_s };
+    }, [&](const SamplerBindingLayout& samplerBinding) -> ResultType {
+        return isSampler(variable, samplerBinding.type) ? ResultType { std::nullopt } : ResultType { "WGSL variable is not a sampler"_s };
+    }, [&](const TextureBindingLayout& textureBinding) -> ResultType {
+        return errorValidatingTexture(variable, textureBinding);
+    }, [&](const StorageTextureBindingLayout& storageTexture) -> ResultType {
+        return isStorageTexture(variable, storageTexture) ? ResultType { std::nullopt } : ResultType { "WGSL variable is not a storage texture"_s };
+    }, [&](const ExternalTextureBindingLayout&) -> ResultType {
+        return isExternalTexture(variable) ? ResultType { std::nullopt } : ResultType { "WGSL variable is not an external texture"_s };
     });
 }
 
-static bool variableAndEntryMatch(const AST::Variable& variable, const BindGroupLayoutEntry& entry)
+static std::optional<String> errorValidatingVariableAndEntryMatch(const AST::Variable& variable, const BindGroupLayoutEntry& entry)
```

Comment:
Does the empty String work for `std::nullopt` case?

Reply:
Yes, assuming I change line 2022:
```
if (auto error = errorValidatingVariableAndEntryMatch(*variable, *entryIt->value))
```
to something like:
```
if (auto error = errorValidatingVariableAndEntryMatch(*variable, *entryIt->value); error.length())
```